### ISSUE TITLE
Abstract protocol attributes cannot be concretised

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#3352]
**Describe the bug**
A protocol attribute whose type is an abstract container cannot receive a concrete type as a value in some circumstances.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That no errors would be produced.

**Screenshots or Code**

```python
from collections.abc import Mapping
from typing import Protocol

class FooType(Protocol):
    foo: Mapping[str, str]


class Foo:
    foo = {'a': 'b'}


class Bar(FooType):
    foo = {'a': 'b'}        # Ok


foo_class: FooType = Foo    # Error
foo: FooType = Foo()        # Error
bar_class: FooType = Bar    # Error
bar: FooType = Bar()        # Ok

```

**VS Code extension or command-line**
Either, v1.1.237

**Comment by erictraut** [comment]
The current behavior is correct. You can verify that mypy emits the same errors in this case.

Because `foo` is a mutable class variable defined in the `FooType` protocol, it is considered invariant. That means all classes that conform to the protocol must use the exact same type for `foo`, not a subtype.

To do this, you need to specify a type annotation in the subclass that matches the type in the protocol class.

```python
class Foo:
    foo: Mapping[str, str] = {"a": "b"}

class Bar(FooType):
    foo: Mapping[str, str] = {"a": "b"}
```

**Comment by layday** [comment]
That being the case, shouldn't 

```py

class Bar(FooType):
    foo = {'a': 'b'} 
```

and 

```py
bar: FooType = Bar()
```

also produce an error?

**Comment by erictraut** [comment]
`Bar` explicitly inherits from `FooType`, so its class variable `foo` has a declared type by inheritance. By contrast, `Foo` Doesn't inherit from any classes, so `foo` has no declared type and is therefore inferred from assignments within the class implementation. Its inferred type is `dict[str, str]`.

**Comment by layday** [comment]
I see, thank you for the clarification.
